File: src/Scripting/Perl/builder.pl
use strict;
use warnings;

sub text_builder {
    my @parts;
    return {
        reset => sub { @parts = (); },
        add_title => sub { push @parts, '# ' . $_[0]; },
        add_section => sub { push @parts, '## ' . $_[0], $_[1]; },
        build => sub { join "\n", @parts },
    };
}

sub html_builder {
    my @parts;
    return {
        reset => sub { @parts = (); },
        add_title => sub { push @parts, '<h1>' . $_[0] . '</h1>'; },
        add_section => sub { push @parts, '<h2>' . $_[0] . '</h2>', '<p>' . $_[1] . '</p>'; },
        build => sub { join '', @parts },
    };
}

sub build_availability_report {
    my ($builder) = @_;
    $builder->{reset}->();
    $builder->{add_title}->('Service status');
    $builder->{add_section}->('Availability', '99.95%');
    return $builder->{build}->();
}

print build_availability_report(text_builder()), "\n---\n";
print build_availability_report(html_builder()), "\n";

File: src/Scripting/Lua/builder.lua
local function text_builder()
  local parts = {}
  return {
    reset = function() parts = {} end,
    add_title = function(title) table.insert(parts, "# " .. title) end,
    add_section = function(heading, body)
      table.insert(parts, "## " .. heading)
      table.insert(parts, body)
    end,
    build = function() return table.concat(parts, "\n") end,
  }
end

local function html_builder()
  local parts = {}
  return {
    reset = function() parts = {} end,
    add_title = function(title) table.insert(parts, "<h1>" .. title .. "</h1>") end,
    add_section = function(heading, body)
      table.insert(parts, "<h2>" .. heading .. "</h2>")
      table.insert(parts, "<p>" .. body .. "</p>")
    end,
    build = function() return table.concat(parts) end,
  }
end

local function build_availability_report(builder)
  builder.reset()
  builder.add_title("Service status")
  builder.add_section("Availability", "99.95%")
  return builder.build()
end

print(build_availability_report(text_builder()))
print("---")
print(build_availability_report(html_builder()))
